src/core/service.c

```diff
@@ -1450,20 +1460,26 @@ static bool service_shall_restart(Service *s) {
 
 static void service_enter_dead(Service *s, ServiceResult f, bool allow_restart) {
         int r;
+        bool shall_restart;
+
         assert(s);
 
         if (s->result == SERVICE_SUCCESS)
                 s->result = f;
 
+        shall_restart = allow_restart && service_shall_restart(s);
+        if (shall_restart)
+                /* Do not close stored fds in the call to service_set_state below */
+                s->preserve_fds = true;
+
         service_set_state(s, s->result != SERVICE_SUCCESS ? SERVICE_FAILED : SERVICE_DEAD);
```

Comment:
i take this back. the patch is fine in this regard. we probably should go through the FAILED/DEAD state here after all